lib/ClangImporter/ClangClassTemplateNamePrinter.cpp

```diff
@@ -59,9 +59,37 @@ struct TemplateInstantiationNamePrinter
         namedArg = typeDefDecl;
       llvm::SmallString<128> storage;
       llvm::raw_svector_ostream buffer(storage);
-      nameImporter->importName(namedArg, version, clang::DeclarationName())
-          .getDeclName()
-          .print(buffer);
+
+      // Print the fully-qualified type name.
+      std::stack<DeclName> qualifiedNameComponents;
+      auto unqualifiedName = nameImporter->importName(namedArg, version);
+      qualifiedNameComponents.push(unqualifiedName.getDeclName());
+      const clang::DeclContext *parentCtx =
+          unqualifiedName.getEffectiveContext().getAsDeclContext();
+      while (parentCtx) {
+        if (auto namedParentDecl = dyn_cast<clang::NamedDecl>(parentCtx)) {
+          // If this component of the fully-qualified name is a decl that is
+          // imported into Swift, remember its name.
+          auto componentName =
+              nameImporter->importName(namedParentDecl, version);
+          qualifiedNameComponents.push(componentName.getDeclName());
+          parentCtx = componentName.getEffectiveContext().getAsDeclContext();
+        } else {
+          // If this component is not imported into Swift, skip it.
+          parentCtx = parentCtx->getParent();
+        }
+      }
+
+      bool needsDot = false;
+      while (!qualifiedNameComponents.empty()) {
```

Comment:
Alternatively we could use [llvm::interleave](https://llvm.org/doxygen/namespacellvm.html#a2145da5bde7663d745e9c3ade392809f) with back iterators.

Reply:
Good point! Replaced `std::stack` with a `std::vector` to be able to use `interleave`.